typing/typemod.ml

```diff
@@ -1957,6 +1957,7 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
           in
           Named (id, param, mty), Types.Named (id, mty.mty_type), newenv, true
       in
+      let newenv = Env.add_lock Alloc_mode.heap newenv in
```

Comment:
Thinking about it some more, this looks like an attempt to prevent local allocations leaking out through functors defined inside expressions that make local allocations.

This could be made to work, but keeping modes correct throughout the module language will be tricky and I don't think it's worthwhile. Instead, I've been trying to ensure that the environment is always fully locked (no accessible locals) when typing anything in the module language.

So I think there is indeed a missing lock, but not here: there should be a lock in Pexp_letmodule where we switch to the module language and nothing in typemod.

Reply:
I don't think we need to be that restrictive. The rules for the module language are easy:
1. structs are all on the heap, when you put values into a struct you need to check that the mode is heap
2. functors are closures on the heap, so the context needs to be locked for their bodies
3. Since all modules are on the heap, unpacking a first-class module must check that its mode is heap

Generally, you only need to lock when you have expressions whose execution is not immediate -- i.e. they are going to be in a closure. Otherwise we only need to ensure that the expected modes are correct.